librz/util/sys.c

```diff
@@ -1826,3 +1791,170 @@ RZ_API int rz_sys_fork(void) {
+#endif
+}
+
+RZ_API int rz_sys_truncate(const char *file, int sz) {
+#if __WINDOWS__
+	int fd = rz_sys_open (file, O_RDWR, 0644);
+	if (fd == -1) {
+		return false;
+	}
+	int r = rz_sys_truncate_fd (fd, sz);
+	if (r != 0) {
+		eprintf ("Could not resize '%s' file\n", file);
+		close (fd);
+		return false;
+	}
+	close (fd);
+	return true;
+#else
+	return truncate (file, sz) == 0;
+#endif
+}
+
+
+/* perm <-> mode */
+RZ_API int rz_sys_open(const char *path, int perm, int mode) {
+	rz_return_val_if_fail (path, -1);
+	char *epath = expand_home (path);
+	int ret = -1;
+#if __WINDOWS__
+	if (!strcmp (path, "/dev/null")) {
+		path = "NUL";
+	}
+	{
+		DWORD flags = 0;
+		if (perm & O_RANDOM) {
+			flags = FILE_FLAG_RANDOM_ACCESS;
+		} else if (perm & O_SEQUENTIAL) {
+			flags = FILE_FLAG_SEQUENTIAL_SCAN;
+		}
+		if (perm & O_TEMPORARY) {
+			flags |= FILE_FLAG_DELETE_ON_CLOSE | FILE_ATTRIBUTE_TEMPORARY;
+		} else if (perm & _O_SHORT_LIVED) {
+			flags |= FILE_ATTRIBUTE_TEMPORARY;
+		} else {
+			flags |= FILE_ATTRIBUTE_NORMAL;
+		}
+		DWORD creation = 0;
+		bool read_only = false;
+		if (perm & O_CREAT) {
+			if (perm & O_EXCL) {
+				creation = CREATE_NEW;
+			} else {
+				creation = OPEN_ALWAYS;
+			}
+			if (mode & S_IREAD && !(mode & S_IWRITE)) {
+				flags = FILE_ATTRIBUTE_READONLY;
+				read_only = true;
+			}
+		} else if (perm & O_TRUNC) {
+			creation = TRUNCATE_EXISTING;
+		}
+		if (!creation || !strcasecmp ("NUL", path)) {
+			creation = OPEN_EXISTING;
+		}
+		DWORD permission = 0;
+		if (perm & O_WRONLY) {
+			permission = GENERIC_WRITE;
+		} else if (perm & O_RDWR) {
+			permission = GENERIC_WRITE | GENERIC_READ;
+		} else {
+			permission = GENERIC_READ;
+		}
+		if (perm & O_APPEND) {
+			permission |= FILE_APPEND_DATA;
+		}
+
+		wchar_t *wepath = rz_utf8_to_utf16 (epath);
+		if (!wepath) {
+			free (epath);
+			return -1;
+		}
+		HANDLE h = CreateFileW (wepath, permission, FILE_SHARE_READ | (read_only ? 0 : FILE_SHARE_WRITE), NULL, creation, flags, NULL);
+		if (h != INVALID_HANDLE_VALUE) {
+			ret = _open_osfhandle ((intptr_t)h, perm);
+		}
+		free (wepath);
+	}
+#else // __WINDOWS__
+	ret = open (epath, perm, mode);
+#endif // __WINDOWS__
+	free (epath);
+	return ret;
+}
+
+RZ_API FILE *rz_sys_fopen (const char *path, const char *mode) {
+	rz_return_val_if_fail (path && mode, NULL);
+	FILE *ret = NULL;
+	char *epath = NULL;
+	if (!epath) {
+		epath = expand_home (path);
+	}
+	if ((strchr (mode, 'w') || strchr (mode, 'a') || rz_file_is_regular (epath))) {
+#if __WINDOWS__
+		wchar_t *wepath = rz_utf8_to_utf16 (epath);
+		if (!wepath) {
+			free (epath);
+			return ret;
+		}
+		wchar_t *wmode = rz_utf8_to_utf16 (mode);
+		if (!wmode) {
+			free (wepath);
+			free (epath);
+			return ret;
+		}
+		ret = _wfopen (wepath, wmode);
+		free (wmode);
+		free (wepath);
+#else // __WINDOWS__
+		ret = fopen (epath, mode);
+#endif // __WINDOWS__
+	}
+	free (epath);
+	return ret;
+}
+
+RZ_API int rz_sys_kill(int pid, int sig) {
```

Comment:
these things are copied-pasted from sandbox.c right?
I think it would be better to check the presence of `kill` in meson.build and base the implementation of these things on `HAVE_KILL` instead of having know ourselves what platforms/compiler support a given function.

Reply:
Yeah, they are 1:1 copies of what was left in sandbox after removing all pure stubs. The approach sounds good to me but I would rather do that in another pr.

Reply:
absolutely.